File: 110-balanced-binary-tree/110-balanced-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean isBalanced(TreeNode root) {
        if(root==null) return true;
        if(root.left==null && root.right==null) return true;
        if((isBalanced(root.left) && isBalanced(root.right))&&((height(root.left)-height(root.right)==1) ||(height(root.left)-height(root.right))==-1 || (height(root.left)==height(root.right)))){
            return true;
        }
        else return false;
    }
    public int height(TreeNode root){
        if(root==null) return -1;
        if(root.left==null && root.right==null) return 0;
        if(root.left!=null){
            if(root.right==null){
                return 1+height(root.left);
            }
            else{
                return 1+Math.max(height(root.left),height(root.right));
            }
        }

                return 1+height(root.right);
    }
}

File: 110-balanced-binary-tree/110-balanced-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    unordered_map<TreeNode*,int> mp;
    bool isBalanced(TreeNode* root) {
        if(root==NULL) return true;
        if(isBalanced(root->left) && isBalanced(root->right) && abs(height(root->left)-               height(root->right))<=1) 
            return true;
        return false;
        
    }
    int height(TreeNode* root){
        if(mp.find(root)!=mp.end()) return mp[root];
        if(root==0){
            mp.insert({root,-1});
            return -1;
        } 
        if(root->left==NULL && root->right==NULL){
            mp.insert({root,0});
            return 0;
        }
        int h=max(height(root->left),height(root->right))+1;
        mp.insert({root,h});
        return h;
        
    }
};